From: Sven Eckelmann <sven.eckelmann@open-mesh.com>
Date: Wed, 25 Jan 2017 10:03:30 +0100
Subject: alfred: Add sha256 PKG_HASH for LEDE

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

Origin: upstream, https://github.com/openwrt-routing/packages/commit/c1c1e134d6a45715eb105e27b74d7fda81176808

diff --git a/alfred/Makefile b/alfred/Makefile
index b5d37ceef6e291d940dfcdf17c91c52b81ccfe07..379c0fe4d84d40723037b05c81aa155d78c50b3a 100644
--- a/alfred/Makefile
+++ b/alfred/Makefile
@@ -14,6 +14,7 @@ PKG_NAME:=alfred
 PKG_VERSION:=2016.5
 PKG_RELEASE:=0
 PKG_MD5SUM:=e03d422ed3b5a162b90e8af13389523f
+PKG_HASH:=37b3babf7f37643cf296be11fb82d5730cf441a5a56f72fba96edae9f149c9d2
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
 PKG_SOURCE_URL:=https://downloads.open-mesh.org/batman/releases/batman-adv-$(PKG_VERSION)
