From: Sven Eckelmann <sven@narfation.org>
Date: Wed, 22 Nov 2017 10:07:07 +0100
Subject: tunneldigger: Update client to fix non-unique L2TP session ids

diff --git a/net/tunneldigger/Makefile b/net/tunneldigger/Makefile
index a57873d38ac9525b2c892885ff03da4b9c2e613a..1144c7f7be91aff737ea3bbf1bb993584172052e 100644
--- a/net/tunneldigger/Makefile
+++ b/net/tunneldigger/Makefile
@@ -2,8 +2,8 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=tunneldigger
 PKG_VERSION:=0.4.7
-PKG_RELEASE:=1
-PKG_REV:=7cc92020cbafe5be2b24eb6bc943a65f151c2a18
+PKG_RELEASE:=2
+PKG_REV:=0a4b3309408254bcf4368acd26e0a1ed4b23363b
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
 PKG_SOURCE_URL:=git://github.com/wlanslovenija/tunneldigger.git
